{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2befc87-0dff-4405-91b0-e248558f3bce",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def wrong_add_function(arg1,arg2):\n",
    "    '''\n",
    "    The function takes in two lists of integers, then it adds\n",
    "    all of arg2 to each item of arg1.\n",
    "    \n",
    "    Example:\n",
    "    > wrong_add_function([1,2,3],[1,1,1])\n",
    "    > [6,9,12]\n",
    "    \n",
    "    whereas the expected correct answer is, [2,3,4]\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    arg1 : list\n",
    "    list of integers.\n",
    "    arg2 : list\n",
    "    list of integers.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    arg1 : list\n",
    "    Elements of arg1, with each element having had the contents of \n",
    "    arg2 added to it.\n",
    "    \n",
    "    '''\n",
    "    arg1_index=0\n",
    "    # NOTE: The expected output [2,3,4] for the input ([1,2,3],[1,1,1])\n",
    "    # implies elementwise addition (arg1[i] + arg2[i]), not adding\n",
    "    # every element in arg2 to each element of arg1.\n",
    "    expected_output = [arg1[i] + arg2[i] for i in range(len(arg1))]\n",
    "    print(f\"The correct answer is supposed to be: {expected_output}. This was generated using a list comprehension and is a lot less code than a while loop.\\n\\n\\n\")\n",
    "    \n",
    "    while arg1_index < len(arg1):\n",
    "        arg_2_sum = 0\n",
    "        # ------------------------------------------------------------------\n",
    "        # We are making an error in the loop!\n",
    "        # This line calculates the sum of (arg1[arg1_index] + all of arg2),\n",
    "        # which is why the output is [6, 9, 12] instead of [2, 3, 4].\n",
    "        # ------------------------------------------------------------------\n",
    "        for arg2_elements in arg2:\n",
    "            print(\"1.a: Here is the error. The sum(arg1[arg1_index]+i for i in arg2]) is adding each value of arg2 to the value at arg1[arg1_index] then summing the iterable together.\")\n",
    "            # ie arg1[0] (1) + arg2[0] (1) = 2, arg1[0] (1) + arg2[1] (1) = 2, arg1[0] (1) + arg2[2] (1) = 2; 2+2+2 = 6\n",
    "            # ie arg1[1] (2) + arg2[0] (1) = 3, arg1[1] (2) + arg2[1] (1) = 3, arg1[1] (2) + arg2[2] (1) = 3; 3+3+3 = 9\n",
    "            # ie arg1[2] (3) + arg2[0] (1) = 4, arg1[2] (3) + arg2[1] (1) = 4, arg1[1] (3) + arg2[3] (1) = 4; 4+4+4 = 12\n",
    "            arg_2_sum = sum([arg1[arg1_index]+i for i in arg2])\n",
    "            print(f\"arg_2_sum = {arg_2_sum}\")\n",
    "        arg1[arg1_index]=arg_2_sum  \n",
    "        arg1_index+=1\n",
    "    print(f\"Finished with wrong_add_function: arg1={arg1}\")\n",
    "    return arg1\n",
    "\n",
    "\n",
    "def correct_add_function(arg1,arg2):\n",
    "    '''\n",
    "    The function takes in two lists of integers, then it adds\n",
    "    all of arg2 to each item of arg1.\n",
    "    \n",
    "    Example:\n",
    "      > wrong_add_function([1,2,3],[1,1,1])\n",
    "      > [6,9,12]\n",
    "    \n",
    "    whereas the expected correct answer is, [2,3,4]\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    arg1 : list\n",
    "      list of integers.\n",
    "    arg2 : list\n",
    "      list of integers.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    arg1 : list\n",
    "      Elements of arg1, with each element having had the contents of \n",
    "      arg2 added to it.\n",
    "    \n",
    "    '''\n",
    "    arg1_index=0\n",
    "\n",
    "    while arg1_index < len(arg1):\n",
    "        # ---We only need to add the element at the same index----\n",
    "        # ---This one line replaces the incorrect inner for loop logic---\n",
    "        arg1[arg1_index] = arg1[arg1_index] + arg2[arg1_index]\n",
    "        arg1_index+=1\n",
    "\n",
    "    # ******************************\n",
    "    # or even more concise would be to remove the outer while loop altogether and just use this:\n",
    "        # expected_output = [arg1[i] + arg2[i] for i in range(len(arg1))]\n",
    "        # print(f\"The correct answer is supposed to be: {expected_output}. This was generated using a list comprehension and is a lot less code than a while loop.\\n\\n\\n\")\n",
    "    # ******************************\n",
    "    \n",
    "    return arg1\n",
    "\n",
    "arg1 = [1,2,3]\n",
    "arg2 = [1,1,1]\n",
    "\n",
    "wrong_add_function(arg1, arg2)\n",
    "\n",
    "arg1 = [1,2,3]\n",
    "arg2 = [1,1,1]\n",
    "correct_add_function(arg1, arg2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1fd4a502-91a0-4376-9372-2f6458ecbb43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[8, 5, 6]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%% try, except\n",
    "'''\n",
    "2.a\n",
    "Update the numeric section of the function with your changes from 1 for both \n",
    "2.b and 2.c\n",
    "\n",
    "2.b\n",
    "Without modifying the string section code itself or the input directly, \n",
    "write a try, except block that catches the issue with the input below and \n",
    "returns an error message to the user, in case users give invalid inputs,\n",
    "(for example an input of [\"5\",\"2\", 5])\n",
    ": \"Your input argument [1 or 2] at element [n]\n",
    "is not of the expected type. Please change this and rerun. Name this function \n",
    "exception_add_function()\n",
    "\n",
    "2.c\n",
    "Without modifying the string section code itself or the input directly, \n",
    "write a try, except block that catches the issue with the input below and \n",
    "gets it to process via the string section. IE, do not, outside the function,\n",
    "change the values of arg_str_1 or arg_str_2. Name this function \n",
    "correction_add_function(), i.e you will not be updating the wrong_add_function,\n",
    "you will simply handle the error of wrong inputs in a seperate function, you want\n",
    "the wrong_add_function to output its current result you are only bolstering the \n",
    "function for edge cases .\n",
    "'''\n",
    "def q2_wrong_add_function(arg1,arg2):\n",
    "   '''\n",
    "   The function takes in two lists of integers, then it adds\n",
    "   all of arg2 to each item of arg1.\n",
    "   \n",
    "   Example:\n",
    "      > wrong_add_function([1,2,3],[1,1,1])\n",
    "      > [4,5,6]\n",
    "   \n",
    "   If the lists are lists of strings, concatenate them\n",
    "   Example:\n",
    "      > wrong_add_function(['1','2','3'],['1','1','1'])\n",
    "      > ['1111','2111','3111']\n",
    "   Parameters\n",
    "   ----------\n",
    "   arg1 : list\n",
    "      list of integers.\n",
    "   arg2 : list\n",
    "      list of integers.\n",
    "\n",
    "   Returns\n",
    "   -------\n",
    "   arg1 : list\n",
    "      Elements of arg1, with each element having had the contents of \n",
    "      arg2 added to it.\n",
    "\n",
    "   '''\n",
    "   #numeric section\n",
    "   if sum([type(i)==int for i in arg1])==len(arg1) and \\\n",
    "      sum([type(i)==int for i in arg2])==len(arg2):\n",
    "         arg1_index=0\n",
    "         while arg1_index < len(arg1):\n",
    "            # if I were to change the code without the \"little as possible\" constraint it would look like this:\n",
    "            # ---We only need to add the element at the same index----\n",
    "            # ---This one line replaces the incorrect inner for loop logic---\n",
    "            arg1[arg1_index] = arg1[arg1_index] + sum(arg2)\n",
    "            arg1_index+=1\n",
    "\n",
    "            # ******************************\n",
    "            # or even more concise would be to remove the outer while loop altogether and just use this:\n",
    "                # expected_output = [arg1[i] + arg2[i] for i in range(len(arg1))]\n",
    "                # print(f\"The correct answer is supposed to be: {expected_output}. This was generated using a list comprehension and is a lot less code than a while loop.\\n\\n\\n\")\n",
    "            # ******************************\n",
    "         return arg1\n",
    "   #string section\n",
    "   elif sum([type(i)==str for i in arg1])==len(arg1) and \\\n",
    "      sum([type(i)==str for i in arg2])==len(arg2):\n",
    "         arg1_index=0\n",
    "         while arg1_index < len(arg1):\n",
    "            arg_2_sum = ''\n",
    "            for arg2_elements in arg2:\n",
    "               arg_2_sum += arg2_elements\n",
    "            arg1[arg1_index]=arg1[arg1_index]+str(arg_2_sum)\n",
    "            arg1_index+=1\n",
    "         return arg1\n",
    "\n",
    "def exception_add_function(arg1, arg2):\n",
    "    '''\n",
    "    The function takes in two lists of integers, \n",
    "    first varifies both arg's are lists\n",
    "    second checks to see if the elements in the arg's are integers and then prints what elements are not integers\n",
    "    \n",
    "    If both lists contain only integers the try statement does no raise an exception and calls q2_wrong_add_function(arg1,arg2).\n",
    "       This allows for the q2_wrong_add_function(arg1,arg2) to execute the int portion of its code\n",
    "    Otherwise a TypeError is raised\n",
    "       If the arg's are lists the user is told this and the code returns without executing any more functions.\n",
    "       If there are more than just integers in the arg's the user is notified of the location of these elements\n",
    "           as a warning but the code is allowed to continue executing and calls correction_add_function(arg1, arg2).\n",
    "           \n",
    "    Example:\n",
    "      > exception_add_function([1,2,3],[1,1,1])\n",
    "      > returns q2_wrong_add_function([1,2,3],[1,1,1])\n",
    "    \n",
    "    If the lists are lists have anything other than ints, User is warned but code continues\n",
    "    Example:\n",
    "      > exception_add_function(['1','2','3'],['1','1','1'])\n",
    "      > *****************************************************\n",
    "      > The first argument passed to this function at element 0 is not of the expected type. Please check this and rerun the code.\n",
    "      > *****************************************************\n",
    "      > *****************************************************\n",
    "      > The first argument passed to this function at element 1 is not of the expected type. Please check this and rerun the code.\n",
    "      > *****************************************************\n",
    "      > etc....\n",
    "      > The function then calls correction_add_function(arg1, arg2) \n",
    "      \n",
    "    Parameters\n",
    "    ----------\n",
    "    arg1 : list\n",
    "      list of integers.\n",
    "    arg2 : list\n",
    "      list of integers.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    if either arg's isn't a list nothing is returned and user is informed \"Both variables that are passed as arguments must of type list for q2_wrong_add_function() to execute.\"\n",
    "    \n",
    "    if arg's contain all int's returns q2_wrong_add_function(arg1, arg2)\n",
    "    \n",
    "    if arg's are lists but have anything other than int's in them:\n",
    "        1. User is given warning about where there are elements other than lists within each argument\n",
    "        2. returns correction_add_function(arg1, arg2)\n",
    "    \n",
    "    '''\n",
    "    try:\n",
    "        if type(arg1) != list or type(arg2) != list:\n",
    "            raise TypeError\n",
    "        \n",
    "        #checks to see if all values are integers and if not raises a type error otherwise if all are integers it goes straight to q2_wrong_add_function\n",
    "        if not sum([type(i)==int for i in arg1])==len(arg1) or not sum([type(i)==int for i in arg2])==len(arg2):\n",
    "            raise TypeError\n",
    "        else:\n",
    "            return q2_wrong_add_function(arg1,arg2)\n",
    "    \n",
    "    except TypeError:\n",
    "        if type(arg1) != list or type(arg2) != list:\n",
    "            print(\"Both variables that are passed as arguments must of type list for q2_wrong_add_function() to execute.\")\n",
    "            return \n",
    "             \n",
    "        #specifies where there are elements that are not integers in the first argument\n",
    "        for i, element in enumerate(arg1):\n",
    "            if type(element) != int:\n",
    "                print(\"*****************************************************\")\n",
    "                print(f\"The first argument passed to this function at element {i} is not of the expected type. Please check this and rerun the code.\")\n",
    "                print(\"*****************************************************\")\n",
    "        #specifies where there are elements that are not integers in the second argument\n",
    "        for i, element in enumerate(arg2):\n",
    "            if type(element) != int:\n",
    "                print(\"*****************************************************\")\n",
    "                print(f\"The second argument passed to this function at element {i} is not of the expected type. Please check this and rerun the code.\")   \n",
    "                print(\"*****************************************************\")\n",
    "        return correction_add_function(arg1, arg2)\n",
    "\n",
    "def correction_add_function(arg1, arg2):\n",
    "    '''\n",
    "    The function takes in two lists, \n",
    "    first double checks to make sure the arg's only contain ints\n",
    "    second if there are elements that are other than int's a ValueError is raised\n",
    "\n",
    "    The ValueError is handled by converting all of the elements to strings regardless of types then calls return q2_wrong_add_function(arg_str_1,arg_str_2)\n",
    "        This ensures all elements of the lists are strings and the \"strings portion\" of q2_wrong_add_function(arg1,arg2) is entered and the lists are processed accordingly.\n",
    "           \n",
    "    Example:\n",
    "      > correction_add_function([1,2,3],[1,1,1])\n",
    "      > returns q2_wrong_add_function([1,2,3],[1,1,1])\n",
    "    \n",
    "    If the lists contain anything other than ints, all elements in all lists are converted to strings and then the q2_wrong_add_function(arg1,arg2) is called.\n",
    "    \n",
    "    Example:\n",
    "      > correction_add_function([\"5\",\"2\",3],[1,1,1])\n",
    "      > prints:\n",
    "        The first argument passed at [2] type was a <class 'int'>.\n",
    "        The first argument passed at [2] type is now a <class 'str'>, value = 3\n",
    "        The second argument passed at [0] type was a  <class 'int'>.\n",
    "        The second argument passed at [0] type is now a <class 'str'>, value = 1\n",
    "        The second argument passed at [1] type was a  <class 'int'>.\n",
    "        The second argument passed at [1] type is now a <class 'str'>, value = 1\n",
    "        The second argument passed at [2] type was a  <class 'int'>.\n",
    "        The second argument passed at [2] type is now a <class 'str'>, value = 1\n",
    "\n",
    "      > The function then returns q2_wrong_add_function(arg1,arg2), thus all elements in both arg's have to be strings and the string portion of the q2_wrong_add_function(arg1,arg2) will run.\n",
    "      \n",
    "    Parameters\n",
    "    ----------\n",
    "    arg1 : list\n",
    "      list of integers.\n",
    "    arg2 : list\n",
    "      list of integers.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    q2_wrong_add_function(arg1,arg2)\n",
    "    \n",
    "    '''\n",
    "    try:\n",
    "        #Checks to see if all elements are integers and if so calls q2_wrong_add_function and aborts the rest of the this function\n",
    "        if sum([type(i)==int for i in arg1])==len(arg1) and sum([type(i)==int for i in arg2])==len(arg2):\n",
    "            return q2_wrong_add_function(arg1,arg2)\n",
    "        \n",
    "        #Checks to see if any of the elements are not strings, raises an error if there are and goes to the except statement \n",
    "        if not sum([type(i)==str for i in arg1])==len(arg1) or not sum([type(i)==str for i in arg2])==len(arg2):\n",
    "            raise ValueError\n",
    "            \n",
    "    except ValueError:\n",
    "        #converts any element in arg1 that is not a str to a string\n",
    "        for i, element in enumerate(arg1):\n",
    "            if type(element) != str:\n",
    "                print(f\"The first argument passed at [{i}] type was a {type(arg1[i])}.\")\n",
    "                arg1[i] = str(arg1[i])\n",
    "                print(f\"The first argument passed at [{i}] type is now a {type(arg1[i])}, value = {arg1[i]}\")\n",
    "        \n",
    "        #converts any element in arg2 that is not a str to a string\n",
    "        for i, element in enumerate(arg2):\n",
    "            if type(element) != str:\n",
    "                print(f\"The second argument passed at [{i}] type was a  {type(arg2[i])}.\")\n",
    "                arg2[i] = str(arg2[i])\n",
    "                print(f\"The second argument passed at [{i}] type is now a {type(arg2[i])}, value = {arg2[i]}\")\n",
    "        #after all the integers have been converted q2_wrong_add_function is called so it will enter the \"string section\"\n",
    "        return q2_wrong_add_function(arg1,arg2)\n",
    "\n",
    "arg_str_1=[5,2,3]\n",
    "arg_str_2=[1,1,1]\n",
    "\n",
    "exception_add_function(arg_str_1,arg_str_2)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
